pkg/controllers/oauthclientscontroller/oauthclientscontroller_test.go

```diff
@@ -0,0 +1,533 @@
+	oauthClientIndexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{})
+
+	routeIndexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{})
+	route := &routev1.Route{
+		ObjectMeta: metav1.ObjectMeta{Name: "oauth-openshift", Namespace: "openshift-authentication"},
+		Spec:       routev1.RouteSpec{Host: masterPublicURL},
+		Status: routev1.RouteStatus{
+			Ingress: []routev1.RouteIngress{
+				{
+					Host: masterPublicURL,
+					Conditions: []routev1.RouteIngressCondition{
+						{Type: routev1.RouteAdmitted, Status: corev1.ConditionTrue},
+					},
+				},
+			},
+		},
+	}
+	if err := routeIndexer.Add(route); err != nil {
+		t.Fatalf("got unexpected error while creating test OAuthsClientsController: %v", err)
+		return nil
+	}
+
+	ingressIndexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{})
+	ingress := &configv1.Ingress{
+		ObjectMeta: metav1.ObjectMeta{Name: "cluster"},
+		Spec: configv1.IngressSpec{
+			Domain: "test.com",
+			ComponentRoutes: []configv1.ComponentRouteSpec{
+				{Namespace: "openshift-authentication", Name: "oauth-openshift", Hostname: masterPublicURL},
+			},
+		},
+	}
+	if err := ingressIndexer.Add(ingress); err != nil {
+		t.Fatalf("got unexpected error while creating test ingress indexer: %v", err)
+		return nil
+	}
+
+	return &oauthsClientsController{
+		oauthClientClient: fakeClientset.OauthV1().OAuthClients(),
+		oauthClientLister: oauthv1listers.NewOAuthClientLister(oauthClientIndexer),
+		routeLister:       routev1listers.NewRouteLister(routeIndexer),
+		ingressLister:     configv1listers.NewIngressLister(ingressIndexer),
+	}
+}
+
+func Test_sync(t *testing.T) {
+	ctx := context.TODO()
+	syncCtx := &fakeSyncContext{}
+	c := newTestOAuthsClientsController(t)
+
+	t.Run("sync-success-non-empty-hostname", func(t *testing.T) {
+		if err := c.sync(ctx, syncCtx); err != nil {
+			t.Errorf("got unexpected err: %v", err)
+		}
+	})
+
+	t.Run("sync-success-empty-hostname", func(t *testing.T) {
+		ingressIndexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{})
+		ingress := &configv1.Ingress{
+			ObjectMeta: metav1.ObjectMeta{Name: "cluster"},
+			Spec: configv1.IngressSpec{
+				Domain:          "test.com",
+				ComponentRoutes: []configv1.ComponentRouteSpec{},
+			},
+		}
+		if err := ingressIndexer.Add(ingress); err != nil {
+			t.Errorf("got unexpected err when setting up test ingress: %v", err)
+		}
+		c.ingressLister = configv1listers.NewIngressLister(ingressIndexer)
+
+		if err := c.sync(ctx, syncCtx); err != nil {
+			t.Errorf("got unexpected err: %v", err)
+		}
+	})
+
+	t.Run("ingress-config-error", func(t *testing.T) {
+		ingressIndexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{})
+		c.ingressLister = configv1listers.NewIngressLister(ingressIndexer)
+
+		if err := c.sync(ctx, syncCtx); err == nil {
+			t.Errorf("expected error but got nil")
+		}
+	})
+
+	t.Run("canonical-route-host-error", func(t *testing.T) {
```

Comment:
you can use table-driven tests here for a more dense and more readable unit test.

Reply:
Updated in e6bf6dd0902c4799309755b5ca632bec6996b080